osu.Game.Rulesets.Mania/Mods/ManiaModHidden.cs

```diff
@@ -6,24 +6,61 @@
 using osu.Framework.Localisation;
 using osu.Game.Rulesets.Mania.UI;
 using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Game.Rulesets.Scoring;
 
 namespace osu.Game.Rulesets.Mania.Mods
 {
-    public class ManiaModHidden : ManiaModPlayfieldCover
+    public partial class ManiaModHidden : ManiaModWithPlayfieldCover
     {
+        /// <summary>
+        /// osu!stable is referenced to 768px.
+        /// </summary>
+        private const float reference_playfield_height = 768;
+
+        private const float min_coverage = 160 / reference_playfield_height;
+        private const float max_coverage = 400f / reference_playfield_height;
+        private const float coverage_increase_per_combo = 0.5f / reference_playfield_height;
+
         public override LocalisableString Description => @"Keys fade out before you hit them!";
         public override double ScoreMultiplier => 1;
 
-        public override BindableNumber<float> Coverage { get; } = new BindableFloat(0.5f)
+        public override Type[] IncompatibleMods => base.IncompatibleMods.Concat(new[]
         {
-            Precision = 0.1f,
-            MinValue = 0.2f,
-            MaxValue = 0.8f,
-            Default = 0.5f,
-        };
-
-        public override Type[] IncompatibleMods => base.IncompatibleMods.Append(typeof(ManiaModFadeIn)).ToArray();
+            typeof(ManiaModFadeIn),
+            typeof(ManiaModCover)
+        }).ToArray();
 
+        public override BindableNumber<float> Coverage { get; } = new BindableFloat(min_coverage);
         protected override CoverExpandDirection ExpandDirection => CoverExpandDirection.AgainstScroll;
+
+        private readonly BindableInt combo = new BindableInt();
+
+        public override void ApplyToScoreProcessor(ScoreProcessor scoreProcessor)
+        {
+            base.ApplyToScoreProcessor(scoreProcessor);
+
+            combo.UnbindAll();
+            combo.BindTo(scoreProcessor.Combo);
+            combo.BindValueChanged(c => Coverage.Value = Math.Min(max_coverage, min_coverage + c.NewValue * coverage_increase_per_combo), true);
+        }
+
+        protected override PlayfieldCoveringWrapper CreateCover(Drawable content) => new LegacyPlayfieldCover(content);
+
+        private partial class LegacyPlayfieldCover : PlayfieldCoveringWrapper
+        {
+            public LegacyPlayfieldCover(Drawable content)
+                : base(content)
+            {
+            }
+
+            protected override float GetHeight(float coverage)
+            {
+                if (DrawHeight == 0)
+                    return base.GetHeight(coverage);
+
+                return base.GetHeight(coverage) * reference_playfield_height / DrawHeight;
```

Comment:
I'm not sure using `DrawHeight` here is the best idea because it appears to break down on very narrow screens, giving them unfair advantage. See the following screenshots:

| | |
| :-: | :-: |
| ![osu_2024-02-12_13-51-48](https://github.com/ppy/osu/assets/20418176/12cd815a-f82b-4d12-a2ce-e1ceb8316703) | ![osu_2024-02-12_13-51-54](https://github.com/ppy/osu/assets/20418176/f0235240-65cd-42fe-94e5-e7caf0c6621d) |

This is the same instant on the same map, same scroll speed, just different aspect ratios. For better comparison here's them next to each other, lined up to the hit position:

![image](https://github.com/ppy/osu/assets/20418176/ff7de2be-23a0-4e54-b268-98029d6e61a8)

The time range is the same, all objects line up to each other, but the effective coverage on the narrow window is considerably less.